lib/websocket/frame.js

```diff
@@ -0,0 +1,223 @@
+   */
+  get terminated () {
+    return this.#fragmentComplete
+  }
+
+  set terminated (value) {
+    this.#fragmentComplete = value
+  }
+
+  /*
+  https://www.rfc-editor.org/rfc/rfc6455#section-5.2
+  0                   1                   2                   3
+  0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+ +-+-+-+-+-------+-+-------------+-------------------------------+
+ |F|R|R|R| opcode|M| Payload len |    Extended payload length    |
+ |I|S|S|S|  (4)  |A|     (7)     |             (16/64)           |
+ |N|V|V|V|       |S|             |   (if payload len==126/127)   |
+ | |1|2|3|       |K|             |                               |
+ +-+-+-+-+-------+-+-------------+ - - - - - - - - - - - - - - - +
+ |     Extended payload length continued, if payload len == 127  |
+ + - - - - - - - - - - - - - - - +-------------------------------+
+ |                               |Masking-key, if MASK set to 1  |
+ +-------------------------------+-------------------------------+
+ | Masking-key (continued)       |          Payload Data         |
+ +-------------------------------- - - - - - - - - - - - - - - - +
+ :                     Payload Data continued ...                :
+ + - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - +
+ |                     Payload Data continued ...                |
+ +---------------------------------------------------------------+
+ */
+  constructor ({
+    opcode = opcodes.TEXT,
+    fin = false,
+    rsv1 = false,
+    rsv2 = false,
+    rsv3 = false,
+    payloadLength,
+    data
+  } = {}) {
+    this.fin = fin
+    this.rsv1 = rsv1
+    this.rsv2 = rsv2
+    this.rsv3 = rsv3
+    this.opcode = opcode
+
+    this.#buffers = [data]
+    this.#offset += data.length
+
+    this.payloadLength = payloadLength
+    this.fragmented = !this.fin && this.opcode !== opcodes.CONTINUATION
+  }
+
+  get data () {
+    return Buffer.concat(this.#buffers, this.#offset)
+  }
+
+  get dataOffset () {
+    return this.#offset
+  }
+
+  addFrame (buffer) {
+    this.#buffers.push(buffer)
+    this.#offset += buffer.length
+  }
+
+  parseCloseBody (onlyCode) {
+    // https://datatracker.ietf.org/doc/html/rfc6455#section-7.1.5
+    /** @type {number|undefined} */
+    let code
+
+    const { data } = this
+
+    if (data.length > 2) {
+      // _The WebSocket Connection Close Code_ is
+      // defined as the status code (Section 7.4) contained in the first Close
+      // control frame received by the application
+      code = data.readUInt16BE(0)
+    }
+
+    if (onlyCode) {
+      if (!isValidStatusCode(code)) {
+        return null
+      }
+
+      return { code }
+    }
+
+    // https://datatracker.ietf.org/doc/html/rfc6455#section-7.1.6
+    /** @type {Buffer} */
+    let reason = data.subarray(2)
+
+    // Remove BOM
+    if (reason[0] === 0xEF && reason[1] === 0xBB && reason[2] === 0xBF) {
+      reason = reason.subarray(3)
```

Comment:
Faster subarray:

```
const FastBuffer = Buffer[Symbol.species]

reason = new FastBuffer(reason.buffer, 3)

```

Reply:
It does not seem to be much faster:

```
$ cat index.js
'use strict';

const benchmark = require('benchmark');

const FastBuffer = Buffer[Symbol.species];
const buf = Buffer.alloc(200).fill('a');

const suite = new benchmark.Suite();

let res;

suite.add('Buffer#subarray()', function () {
  res = buf.subarray(100);
});

suite.add('FastBuffer', function () {
  res = new FastBuffer(buf.buffer, buf.byteOffset + 100, buf.length - 100);
});

suite.on('cycle', function (event) {
  console.log(String(event.target));
});

suite.on('complete', function () {
  console.log('Fastest is ' + this.filter('fastest').map('name'));
});

suite.run({ async: true });
```

```
$ node -v
v18.12.1
```

```
$ node index.js
Buffer#subarray() x 22,040,974 ops/sec ±2.21% (86 runs sampled)
FastBuffer x 22,715,035 ops/sec ±2.60% (85 runs sampled)
Fastest is FastBuffer
```

Reply:
but it is a little faster :D

Reply:
I will look into all the subarray optimizations in a future pr